src/core/bpffs-program.c

```diff
@@ -0,0 +1,92 @@
+
+#include "bpffs-program.h"
+#include "cgroup-bpf-ctx.h"
+
+DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(bpf_prog_hash_ops, void, trivial_hash_func, trivial_compare_func, BPFProgram, bpf_program_unref);
+
+static int load_programs(Unit *u, const CGroupContext *c, Hashmap *progs_by_id) {
+        CGroupBPFFsProgram *p;
+        int r;
+
+        assert(u);
+        assert(c);
+        assert(progs_by_id);
+
+        LIST_FOREACH(prog, p, c->bpffs_programs) {
+                _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+
+                r = bpf_program_new_from_bpffs_path(p->bpffs_path, p->attach_type, &prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Loading of BPF program %s failed: %m", p->bpffs_path);
+
+                r = hashmap_put(progs_by_id, p, prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't put program to hashmap: %m");
+
+                TAKE_PTR(prog);
+        }
+
+        return 0;
+}
+
+static int attach_programs(Unit *u, const char *path, Hashmap *progs_by_id, Set **set_installed, uint32_t attach_flags) {
+        CGroupBPFFsProgram *id;
+        BPFProgram *prog;
+        int r;
+
+        assert(u);
+        assert(set_installed);
+
+        HASHMAP_FOREACH_KEY(prog, id, progs_by_id) {
+                r = bpf_program_cgroup_attach(prog, id->attach_type, path, attach_flags);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Attaching custom egress BPF program to cgroup %s failed: %m", path);
+
+                r = set_ensure_put(set_installed, &bpf_prog_hash_ops, prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't add program to BPF program set: %m");
+
+                bpf_program_ref(prog);
+        }
+
+        return 0;
+}
+
+int bpffs_program_install(Unit *u) {
+        _cleanup_free_ char *cgroup_path = NULL;
+        /* BPFProgram is identified by CGroupBPFFsProgram key. */
+        _cleanup_hashmap_free_ Hashmap *progs_by_id = NULL;
+        CGroupContext *cc;
+        int r;
+
+        assert(u);
+
+        cc = unit_get_cgroup_context(u);
+        if (!cc)
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_path);
+        if (r < 0)
+                return r;
+
+        if (LIST_IS_EMPTY(cc->bpffs_programs) && set_isempty(u->bpf_custom_installed))
+                return 0;
+
+        progs_by_id = hashmap_new(&bpf_prog_hash_ops);
+        if (!progs_by_id)
+                return log_oom();
+
+        r = load_programs(u, cc, progs_by_id);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Loading programs from bpffs failed: %m");
+
+        /* Programs are detached in bpf_program_free. */
+        set_clear(u->bpf_custom_installed);
+
+        r = attach_programs(u, cgroup_path, progs_by_id, &u->bpf_custom_installed, BPF_F_ALLOW_MULTI);
```

Comment:
I got a couple of questions on this line:

1. How does it play with the bpf-firewall? IIUC the bpf-firewall [doesn't ](https://github.com/systemd/systemd/blob/fee6441601c979165ebcbb35472036439f8dad5f/src/core/bpf-firewall.c#L701)always use the `BPF_F_ALLOW_MULTI` flag. For instance, with the following unit file "myprogram" is never executed:

```
[Unit]
Description=My test unit
[Service]
ExecStart=-/bin/sh -c 'ping 8.8.8.8'
IPAddressDeny=9.9.9.9
BPFProgram=egress:/sys/fs/bpf/myprogram
```
There is some [logic](https://github.com/systemd/systemd/blob/fee6441601c979165ebcbb35472036439f8dad5f/src/core/bpf-firewall.c#L710) to enable that flag when `IP{Ingress,Egress}FilterPath` is used. Should be there a similar logic for the properties introduced on this PR? 

@poettering I noticed you [introduced](https://github.com/systemd/systemd/commit/acf7f253dedbca7ba5edada37d11221a47101ee3#diff-9acac02f0788de0b9791ee107913ea05fc8d739b3b4ad46a23f8775370c66f75R589) the logic to use `BPF_F_ALLOW_MULTI` some years ago, do you have any comments here? 
I'm asking because in https://github.com/systemd/systemd/pull/18385 I have the same problem, bpf-firewall doesn't play nice with `RestrictNetworkInterfaces`. Can't we enable `BPF_F_ALLOW_MULTI` when it's supported without taking into consideration cgroup delegation and unit type in the bpf-firewall? 

2. In the bpf-firewall there is some code to test if this flag is supported, that code is not present in this PR. I think it's fine to assume the `BPFProgram` feature is only available in kernels supporting `BPF_F_ALLOW_MULTI`, but I wanted to confirm this is a conscious decision.

Reply:
@mauriciovasquezbernal

Nice catch.
There is a [proposed change](https://github.com/systemd/systemd/pull/17495/commits/4e42210d40f96e185a55d43041dd6b962ea830dd#diff-9acac02f0788de0b9791ee107913ea05fc8d739b3b4ad46a23f8775370c66f75) to use `BPF_F_ALLOW_MULTI` more aggressively, i.e. if kernel supports it. I tweaked `bpf-firewall.c` code in some previous versions of this PR, but didn't touch that in final one having https://github.com/systemd/systemd/pull/17495 in mind.
A similar change was made for `BPF_CGROUP_DEVICE`.
Wrt cgroup skb egress and ingress, I'll actualize `bpf-firewall.c` according to https://github.com/systemd/systemd/pull/17495

Reply:
@mauriciovasquezbernal

Added unconditional `BPF_F_ALLOW_MULTI` to the stack
```
IPAddressDeny=any
IPAddressAllow=localhost
BPFProgram=egress:/sys/fs/bpf/trivial/cgroup_skb_egress
ExecStart=-/bin/ping 127.0.0.1 -W 1
```
Journal logs:
```
Feb 05 02:19:11 image systemd[1]: Started BPF foreign test.
Feb 05 02:19:11 image ping[103]: PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
Feb 05 02:19:11 image ping[103]: /bin/ping: sendmsg: Operation not permitted
Feb 05 02:19:12 image ping[103]: /bin/ping: sendmsg: Operation not permitted
```
bpftool c t:
```
/sys/fs/cgroup/system.slice/_bpf-foreign.service
    878      ingress         multi                          
    877      egress          multi                          
    876      egress          multi           trivial_egress
```